src/form-detection.js

```diff
@@ -1,3 +1,249 @@
+  const elementFromID = findElm(type);
+  if (elementFromID.length) {
+    return elementFromID;
+  }
+
+  // jQuery selectors are case sensitive so this function will add all case variations.
+  // We do this instead of overriding jQuery's selection method in case it's being used in other
+  // scripts on the page.
+  const expandLabels = labels =>
+    labels.reduce((acc, label) => [...acc, label, label.toLowerCase(), label.toUpperCase()], []);
+
+  const potentialLabels = {
+    primary: expandLabels(['Primary', 'Address']),
+    secondary: expandLabels(['Address 2', 'Secondary', 'Apartment', 'Suite']),
+    city: expandLabels(['City']),
+    state: expandLabels(['State']),
+    zip: expandLabels(['Zip', 'Postal']),
+    country: expandLabels(['Country'])
+  };
+
+  /**
+   * Chain together the query selectors of each label. When searching for the primary line we
+   * clarify two variations of an address form:
+   *    Form A           Form B
+   *    Address 1        Address
+   *    Address 2        Apartment, Suite, Unit, etc.
+   * Matching on "Address" would include the secondary label and raise an error, so we exclude
+   * it in the selector.
+   */
+  const selector = potentialLabels[type].map(currentLabel =>
+    type === 'primary' && currentLabel.toLowerCase() === 'address'
+      ? `:contains('${currentLabel}'):not(:contains('${currentLabel} 2'))`
+      : `:contains('${currentLabel}')`
+  ).join(', ');
+
+
+  const selections = $(selector);
+  if (!selections.length) {
+    return null;
+  }
+
+  // Raise an error on multiple matching labels for the user to resolve 
+  if (selections.filter("label").length > 1) {
+    return selections.filter("label");
+  }
+
+  // Selections are ordered from furthest to closest, so the last element most likely contains
+  // the label text itself
+  const label = $(selections[selections.length - 1]);
+  const inputId = label.attr("for");
+
+  if (inputId) {
+    const inputSelections = $(`#${inputId}`);
+    return inputSelections;
+  }
+
+  const siblingSelections = label.siblings("input");
+  if (siblingSelections.length) {
+    return siblingSelections;
+  }
+
+  const childSelections = label.children("input");
+  if (childSelections.length) {
+    return childSelections;
+  }
+
+  const parentSelections = label.parentsUntil("form", "input");
+  if (parentSelections.length) {
+    return parentSelections;
+  }
+
+  return null;
+};
+
+
+const resolveParsingResults = addressElements => {
+  let parseResultError = ''; 
+
+  const missingElements = Object.keys(addressElements).filter(key => {
+    // Omit country input in case form is domestic only
+    if (key === 'country') {
+      return false;
+    }
+
+    const searchResult = addressElements[key];
+    return searchResult === null || (searchResult && searchResult.length === 0);
+  });
+
+  const multipleElements = Object.keys(addressElements).filter(key => {
```

Comment:
Can you account for the element type and/or its display/visibility state (usually display) as a way of ensuring that the duplication is not due to a hidden form element (type='hidden') or due to the CSS display property for the input (or inherited display property on an ancestor) being 'none'. I've seen some address forms with hidden inputs and it would be good to ignore these. We can also add in the next release as a refinement to the process. Fine either way...